src/catchup/CatchupManagerImpl.cpp

```diff
@@ -450,16 +511,47 @@ CatchupManagerImpl::tryApplySyncingLedgers()
         auto const& lcd = it->second;
 
         // we still have a missing ledger
-        if (ledgerHeader.header.ledgerSeq + 1 != lcd.getLedgerSeq())
+        if (nextToClose != lcd.getLedgerSeq())
+        {
+            break;
+        }
+
+        // If we have too many ledgers queued to apply, just stop scheduling
+        // more and let the node gracefully go into catchup.
+        releaseAssert(mLastQueuedToApply >= lcl);
+        if (nextToClose - lcl >= MAX_EXTERNALIZE_LEDGER_APPLY_DRIFT)
         {
+            CLOG_INFO(History,
+                      "Next ledger to apply is {}, but LCL {} is too far "
+                      "behind, waiting",
+                      nextToClose, lcl);
             break;
         }
 
-        mApp.getLedgerManager().closeLedger(lcd);
-        CLOG_INFO(History, "Closed buffered ledger: {}",
-                  LedgerManager::ledgerAbbrev(ledgerHeader));
+        if (mApp.getConfig().parallelLedgerClose())
+        {
+            // Notify LM that application has started
+            mApp.getLedgerManager().beginApply();
+            mApp.postOnLedgerCloseThread(
+                [&app = mApp, lcd]() {
+                    // No-op if app is shutting down
+                    if (app.isStopping())
+                    {
+                        return;
+                    }
+                    app.getLedgerManager().closeLedger(lcd,
+                                                       /* externalize */ true);
+                },
+                "closeLedger queue");
+        }
+        else
+        {
+            mApp.getLedgerManager().closeLedger(lcd, /* externalize */ true);
+        }
+        mLastQueuedToApply = lcd.getLedgerSeq();
```

Comment:
oh wow now I am very confused. I thought mLastQueuedToApply was tracking the _back_ of the queue. Now it looks like it's tracking the _front_.